wallet-api/src/Wallet/UTXO/Index.hs

```diff
@@ -0,0 +1,59 @@
+-- | A map of transaction hashes to transactions for efficient lookups of
+--   values on the mockchain.
+module Wallet.UTXO.Index(
+    BlockchainLookup(..),
+    BlockchainIndex(..),
+    empty,
+    insert,
+    initialise,
+    Lookup,
+    runLookup,
+    LookupError(..)
+    ) where
+
+import           Control.Monad.Except (MonadError (..), liftEither)
+import           Control.Monad.Reader (MonadReader (..), ReaderT (..), ask)
+import           Data.Foldable        (fold, foldl')
+import qualified Data.Map             as Map
+import           Data.Semigroup       (Semigroup)
+import           Prelude              hiding (lookup)
+import           Wallet.UTXO.Types    (Blockchain, Tx, TxId', hashTx)
+
+-- | The ability to find transactions by their hashes
+class BlockchainLookup m where
```

Comment:
I thought about that too. 

The purpose of `BlockchainLookup` is to facilitate the implementation of [mkPending](https://github.com/input-output-hk/plutus/pull/205/commits/faad28ff83ca8c1ef9155fbee5aecd19ad3166e4#diff-98512c8229543d1cea78d5533fbda0ecR122) and in there we don't need anything else from `MonadEmulator`. In fact, we can get by with an `Applicative` instead of `Monad` constraint which I think is quite nice.

That's also why I delegated the error handling to the caller of `mkPending` (re. your comment below). If a single lookup fails then the whole transaction is invalid and there is nothing we can do about it inside `mkPending`.

Maybe I could replace it with `type BlockchainLookup m = (MonadReader BlockchainIndex m, MonadError LookupError m)`?

Reply:
Okay, but I'd like it to be a superclass of `MonadEmulator`, then. You clearly can do lookups if you're in `MonadEmulator`, I think we should expose that.

Reply:
But if you are in `MonadEmulator` then you probably want to handle the lookup error yourself. For example, if this happens while validating a transaction you could reject the transaction.

It's only inside `validTx` (and, in the future, `mkPending`) where we can't recover from that.

Reply:
Hm, that makes me think maybe we should just return `Maybe` then. I might need to fire this up in an editor and have a think about it, but it's not blocking if you're happy with the current design.

Reply:
I'm happy with it, but I'm open to changing it to something better!